SYCL/Basic/code_location_e2e.cpp

```diff
@@ -0,0 +1,147 @@
+// RUN: %clangxx -fsycl -DNDEBUG %s -o %t.out
+// RUN: %CPU_RUN_PLACEHOLDER %t.out %CPU_CHECK_PLACEHOLDER
+// RUN: %GPU_RUN_PLACEHOLDER %t.out %GPU_CHECK_PLACEHOLDER
+// RUN: %ACC_RUN_PLACEHOLDER %t.out %ACC_CHECK_PLACEHOLDER
+
+// RUN: %clangxx -fsycl %s -o %t.out
+// RUN: %CPU_RUN_PLACEHOLDER %t.out %CPU_CHECK_PLACEHOLDER
+// RUN: %GPU_RUN_PLACEHOLDER %t.out %GPU_CHECK_PLACEHOLDER
+// RUN: %ACC_RUN_PLACEHOLDER %t.out %ACC_CHECK_PLACEHOLDER
+
+/*
+    clang++ -fsycl -DNDEBUG -o smyl.bin code_location_e2e.cpp   // <<--- NDEBUG
+   will suppress fileLocation
+
+    -DNDEBUG  <-- not debugging, meaning fileName should not be retrievable (for
+   data privacy reasons)
+
+   Between having to test against NDEBUG and test from kernel code,
+   using assert/abort type of logic is not available. This test just
+   outputs strings that it checks with FileCheck.
+
+   This test uses hardcoded file code location positions. If modified, those
+   may change. pass in any value to signal "report only" and not run tests.
+
+*/
+
+#include <sycl/sycl.hpp>
+using namespace sycl;
+
+// TODO: switch to strcmp
+//  Presently there is a conflict with <cstring> functions and
+//  sycl::stream. This bespoke string comparison avoids issue.
+bool stringsAreSameP(const char *a, const char *b){
+    // if both are nullptr, then they are the same. 
+    if ((a == nullptr) && (b == nullptr))
+      return true;
+    // but if only one, they are not
+    if ((a == nullptr) || (b == nullptr))
+      return false;
+
+    int index = 0;
+    while(true){
+        if(a[index] != b[index]) { return false; }
+        if(a[index] == '\0'){ return true; } // if we are on this line we know a[i]==b[i]
+        index++; 
+    }
+    // we will never arrive here
+    return true;
```

Comment:
```suggestion
    // We will never arrive here.
    return false;
```

IMO, indicating failure in the unreachable case is more logical.

Reply:
yeah, but no good way to do that without complicating the interface. Can't throw an exception in code called by kernels.  return false seems best compromise for code not actually in the end-product.